static/app/components/notificationActions/notificationActionManager.tsx

```diff
@@ -0,0 +1,203 @@
+    // Updates notif action from state using the index
+    const updatedActions = [...notificationActions];
+    updatedActions.splice(index, 1, updatedAction);
+    setNotificationActions(updatedActions);
+  };
+
+  // Lists the available actions for each service
+  const availableServices: Record<
+    NotificationActionService,
+    AvailableNotificationAction[]
+  > = {
+    [NotificationActionService.SENTRY_NOTIFICATION]: [],
+    [NotificationActionService.EMAIL]: [],
+    [NotificationActionService.SLACK]: [],
+    [NotificationActionService.PAGERDUTY]: [],
+    [NotificationActionService.MSTEAMS]: [],
+    [NotificationActionService.SENTRY_APP]: [],
+  };
+  availableActions.forEach(a => {
+    availableServices[a.action.serviceType as NotificationActionService].push(a);
+  });
+
+  // Groups the notification actions together by service
+  // Will render the notif actions in the order the keys are listed in
+  const actionsMap: Record<
+    NotificationActionService,
+    {action: NotificationAction; index: number}[]
+  > = {
+    [NotificationActionService.SENTRY_NOTIFICATION]: [],
+    [NotificationActionService.EMAIL]: [],
+    [NotificationActionService.SLACK]: [],
+    [NotificationActionService.PAGERDUTY]: [],
+    [NotificationActionService.MSTEAMS]: [],
+    [NotificationActionService.SENTRY_APP]: [],
+  };
+  notificationActions.forEach((action, index) => {
+    if (action.serviceType) {
+      actionsMap[action.serviceType].push({action, index});
+    }
+  });
+
+  // Groups the pagerduty integrations with their corresponding allowed services
+  const pagerdutyIntegrations: Record<number, AvailableNotificationAction[]> = {};
+  availableServices[NotificationActionService.PAGERDUTY].forEach(service => {
+    const integrationId = service.action.integrationId;
+    if (integrationId) {
+      if (integrationId in pagerdutyIntegrations) {
+        pagerdutyIntegrations[integrationId].push(service);
+      } else {
+        pagerdutyIntegrations[integrationId] = [service];
+      }
+    }
+  });
+
+  function renderNotificationActions() {
+    if (!notificationActions) {
+      return null;
+    }
+
+    const serviceActions: NotificationAction[] = [];
+
+    // Renders the notif actions grouped together by kind
+    Object.keys(actionsMap).forEach(serviceType => {
+      const services = actionsMap[serviceType];
+      serviceActions.push(
+        services.map(({action, index}) => (
+          <NotificationActionItem
+            key={index}
+            index={index}
+            defaultEdit={!action.id}
+            action={action}
+            recipientRoles={recipientRoles}
+            availableActions={availableServices[serviceType]}
+            pagerdutyIntegrations={pagerdutyIntegrations}
+            project={project}
+            onDelete={removeNotificationAction}
+            onUpdate={updateNotificationAction}
+          />
+        ))
+      );
+    });
+    return serviceActions;
```

Comment:
lets just return the map rather than using forEach to populate a const?

Reply:
forEach is because it's a Record<>